Handle user login attemp on unauthorized panel

underdpt opened the pull request:
Hello,

When a user tries to login on a panel for which he doesn't have permissions, he's shown a 403 error but the user remains logged in.

This PR verifies that the user has access to the panel during the login process, and if the user isn't allowed to access the panel, it logouts the user and show the credentials mismatch validation error.

- [x ] Changes have been thoroughly tested to not break existing functionality.
- [ ] New functionality has been documented or existing documentation has been updated to reflect changes.
There's no documentation about the login process.

- [ ] Visual changes are explained in the PR description using a screenshot/recording of before and after.
There's no visual changes on this PR

underdpt committed:
Handle user login attemp on unauthorized panel

Hello,

When a user tries to login on a panel for which he doesn't have permissions, he's shown a 403 error but the user remains logged in.

This PR verifies that the user has access to the panel during the login process, and if the user isn't allowed to access the panel, it logouts the user and show the credentials mismatch validation error.

underdpt committed:
Handle user login attemp on unauthorized panel (#5)

Hello,

When a user tries to login on a panel for which he doesn't have permissions, he's shown a 403 error but the user remains logged in.

This PR verifies that the user has access to the panel during the login process, and if the user isn't allowed to access the panel, it logouts the user and show the credentials mismatch validation error.

underdpt committed:
chore: fix code style

underdpt committed:
chore: fix code style

danharrin commented on packages/panels/src/Pages/Auth/Login.php:
```diff
@@ -75,6 +75,14 @@ public function authenticate(): ?LoginResponse
+        if (! Filament::auth()->user()->canAccessPanel(Filament::getCurrentPanel())) {
+            Filament::auth()->logout();
+
+            throw ValidationException::withMessages([
```
Maybe let's extract this throw into its own method `protected function throwFailureValidationException(): void` and then reuse it in both places?

underdpt replied:
Done. Returning `never` instead of `void` as we're on PHP8.1.

danharrin commented on packages/panels/src/Pages/Auth/Login.php:
```diff
@@ -75,6 +75,14 @@ public function authenticate(): ?LoginResponse
+        if (! Filament::auth()->user()->canAccessPanel(Filament::getCurrentPanel())) {
+            Filament::auth()->logout();
+
+            throw ValidationException::withMessages([
```
Maybe let's extract this throw into its own method `protected function throwFailureValidationException(): void` and then reuse it in both places?

danharrin commented:
Also PHPstan appears to be failing? And you should add a test for this.

underdpt committed:
Merge branch 'filamentphp:3.x' into main

underdpt committed:
Move throw ValidationException to his own method

underdpt committed:
Fix PHPStan

underdpt committed:
Merge branch 'filamentphp:3.x' into handle-user-login-unauthorized-panel

underdpt committed:
Handle user login unauthorized panel (#6)

* Handle user login attemp on unauthorized panel

Hello,

When a user tries to login on a panel for which he doesn't have permissions, he's shown a 403 error but the user remains logged in.

This PR verifies that the user has access to the panel during the login process, and if the user isn't allowed to access the panel, it logouts the user and show the credentials mismatch validation error.

* chore: fix code style

* Move throw ValidationException to his own method

* Fix PHPStan

underdpt committed:
Test: add custom panel.

NOTE that adding the CustomPanelProvider after AdminPanelProvider would break some tests

underdpt committed:
Test: User can only access "admin" panel.

This is so existing tests are unaffected.

underdpt committed:
Test: it cannot authenticate on unauthorized panel

underdpt committed:
Merge remote-tracking branch 'origin/handle-user-login-unauthorized-panel'

underdpt committed:
Merge branch 'main' of https://github.com/underdpt/filament

underdpt commented:
Hi,

I've submitted all changes.

I don't know why all checks are failing, they all work on my local environment, and it seems related to CI:

```
Error: Can't use 'tar -xzf' extract archive file: /home/runner/work/_actions/_temp_42897cc5-320e-41f8-9537-4c6ce62b604e/bf82fa85-cb66-4e2c-b2af-f15cf547fbe4.tar.gz. return code: 2.
```

wychoong commented:
please review your PR, seems like there are some codes duplicated during merge

underdpt committed:
Update Login.php

Fix duplicate code error

underdpt commented:
@wychoong thanks, i'm sorry that got leaked to the PR!

underdpt committed:
chore: fix code style

danharrin committed:
Update TestCase.php
